streams/fileplayer.py

```diff
@@ -0,0 +1,263 @@
+""" Play a file using vlc """
+
+import os
+import sys
+import time
+import json
+import vlc
+import argparse
+from typing import List, Optional, Any, IO
+
+parser = argparse.ArgumentParser(prog='fileplayer', description='play and control an audio file using vlc')
+parser.add_argument('url', type=str, help='internet radio station url')
+parser.add_argument('output', type=str, help='alsa output', nargs='?', default=None)
+parser.add_argument('--song-info', type=str, help='file to update with current song information in json format')
+parser.add_argument('--log', type=str, help='log file (defaults to stdout)', default=None)
+parser.add_argument('--cmd', type=str, help='command file', default=None)
+parser.add_argument('--test', action='store_true', help='verify the url is valid and return')
+parser.add_argument('--verbose', action='store_true', help='show more verbose output')
+args = parser.parse_args()
+
+config = "--aout=alsa "
+if args.test:
+  config += " --alsa-audio-device null"
+elif args.output:
+  alsa_device = args.output
+  config += " --alsa-audio-device {}".format(alsa_device)
+
+log_file: Optional[IO[Any]] = None
+if args.log:
+  try:
+    os.remove(args.log)
+  except Exception:
+    pass
+  log_file = open(args.log, 'a', encoding='utf-8')
+
+
+def log(info):
+  if log_file:
+    try:
+      print(info, file=log_file)
+      log_file.flush()
+    except:
+      print(f'Error writing to logfile: {args.log}')
+      print(info)
+  else:
+    print(info)
+
+
+def update_info(info) -> bool:
+  try:
+    with open(args.song_info, "wt", encoding='utf-8') as info_file:
+      info_file.write(json.dumps(info))
+    return True
+  except Exception:
+    log('Error: %s' % sys.exc_info()[1])
+    return False
+
+
+log_file: Optional[IO[Any]] = None
+if args.log:
+  try:
+    os.remove(args.log)
+  except Exception:
+    pass
+  log_file = open(args.log, 'a', encoding='utf-8')
+
+
+log("Starting VLC instance.")
+instance = vlc.Instance(config.split())
+try:
+  media = instance.media_new(args.url)
+except (AttributeError, NameError) as e:
+  log('%s: %s (%s LibVLC %s)' % (e.__class__.__name__, e, sys.argv[0], vlc.libvlc_get_version()))
+  sys.exit(1)
+try:
+  player = instance.media_player_new()
+  player.set_media(media)
+  player.play()
+except Exception:
+  log(sys.exc_info())
+  sys.exit(1)
+
+# Keep track of the current state so we only update on change
+cur_url = ''
+cur_info = {
+  'track': '',
+  'artist': '',
+  'station': '',
+  'state': 'stopped',
+}
+if args.song_info:
+  if not update_info(cur_info):
+    sys.exit(1)
+
+restarts: List[float] = []
+
+
+def restart_vlc():
```

Comment:
I'm unsure how useful the restart logic is in this script. We can leave it in for now though